cli/lsp/analysis.rs

```diff
@@ -264,6 +299,248 @@ pub struct CodeLensData {
+      ));
+    }
+    let edit =
+      ts_changes_to_edit(&action.changes, index_provider, version_provider)
+        .await?;
+    let code_action = lsp::CodeAction {
+      title: action.description.clone(),
+      kind: Some(lsp::CodeActionKind::QUICKFIX),
+      diagnostics: Some(vec![diagnostic.clone()]),
+      edit,
+      command: None,
+      is_preferred: None,
+      disabled: None,
+      data: None,
+    };
+    self.actions.retain(|(c, a)| {
+      !(action.fix_name == a.fix_name && code_action.edit == c.edit)
+    });
+    self.actions.push((code_action, action.clone()));
+
+    if let Some(fix_id) = &action.fix_id {
+      if let Some((existing_fix_all, existing_action)) =
+        self.fix_all_actions.get(fix_id)
+      {
+        self.actions.retain(|(c, _)| c != existing_fix_all);
+        self
+          .actions
+          .push((existing_fix_all.clone(), existing_action.clone()));
+      }
+    }
+    Ok(())
+  }
+
+  /// Add a TypeScript action to the actions as a "fix all" action, where it
+  /// will fix all occurrences of the diagnostic in the file.
+  pub async fn add_ts_fix_all_action<F, Fut, V>(
+    &mut self,
+    action: &tsc::CodeFixAction,
+    diagnostic: &lsp::Diagnostic,
+    combined_code_actions: &tsc::CombinedCodeActions,
+    index_provider: &F,
+    version_provider: &V,
+  ) -> Result<(), AnyError>
+  where
+    F: Fn(ModuleSpecifier) -> Fut + Clone,
+    Fut: Future<Output = Result<LineIndex, AnyError>>,
+    V: Fn(ModuleSpecifier) -> Option<i32>,
+  {
+    if combined_code_actions.commands.is_some() {
+      return Err(custom_error(
+        "UnsupportedFix",
+        "The action returned from TypeScript is unsupported.",
+      ));
+    }
+    let edit = ts_changes_to_edit(
+      &combined_code_actions.changes,
+      index_provider,
+      version_provider,
+    )
+    .await?;
+    let title = if let Some(description) = &action.fix_all_description {
+      description.clone()
+    } else {
+      format!("{} (Fix all in file)", action.description)
+    };
+
+    let code_action = lsp::CodeAction {
+      title,
+      kind: Some(lsp::CodeActionKind::QUICKFIX),
+      diagnostics: Some(vec![diagnostic.clone()]),
+      edit,
+      command: None,
+      is_preferred: None,
+      disabled: None,
+      data: None,
+    };
+    if let Some((existing, _)) =
+      self.fix_all_actions.get(&action.fix_id.clone().unwrap())
+    {
+      self.actions.retain(|(c, _)| c != existing);
+    }
+    self.actions.push((code_action.clone(), action.clone()));
+    self.fix_all_actions.insert(
+      action.fix_id.clone().unwrap(),
+      (code_action, action.clone()),
+    );
+    Ok(())
+  }
+
+  /// Move out the code actions and return them as a `CodeActionResponse`.
+  pub fn get_response(self) -> Option<lsp::CodeActionResponse> {
+    let code_actions = self
+      .actions
+      .into_iter()
+      .map(|(c, _)| lsp::CodeActionOrCommand::CodeAction(c))
+      .collect();
+    Some(code_actions)
```

Comment:
Seems like this never can return None? Consider changing return type to just `CodeActionResponse`

Reply:
It does mean I have to wrap it in `Some()` where it get invoked, because the response back requires requires the `Option`.

Reply:
ok, I like it better that way anyways.